lib/HLSL/LiveValueAnalysis.cpp

```diff
@@ -0,0 +1,1323 @@
+    // cut into chunks as it won't handle a massive string.
+    const size_t outputSize = VSStr.str().length();
+    const size_t blockSize = 1024;
+    for (int i = 0; i < outputSize;) {
+      OutputDebugStringA(VSStr.str().substr(i, blockSize).c_str());
+      i += blockSize;
+    }
+
+    PrettyStr.flush();
+    VSStr.flush();
+  }
+
+  // No changes.
+  return false;
+}
+
+template<class T, class C>
+std::vector<T> LiveValueAnalysis::sortInstructions(C& c)
+{
+  std::vector<T> sorted;
+
+  for (T inst : c)
+    sorted.push_back(inst);
+
+  if (sorted.empty())
+    return sorted;
+
+  DominatorTree DT;
+  DT.recalculate(*sorted[0]->getParent()->getParent());
+
+  std::sort(sorted.begin(), sorted.end(), [&] (T a, T b) -> bool {
+    if (DT.dominates(a->getParent(), b->getParent()))
+      return true;
+    else if (DT.dominates(b->getParent(), a->getParent()))
+      return false;
+
+    Function::iterator itA( a->getParent() );
+    Function::iterator itB( b->getParent() );
+    size_t ad = std::distance( a->getParent()->getParent()->begin(), itA );
+    size_t bd = std::distance( a->getParent()->getParent()->begin(), itB );
+
+    return ad < bd;
+  });
+
+  return sorted;
+}
+
+bool LiveValueAnalysis::usesUavResource(CallInst* call) {
+  CallInst* createHandle = dyn_cast<CallInst>(call->getArgOperand(1));
+  assert(createHandle);
+  assert(createHandle->getCalledFunction()->getName().startswith("dx.op.createHandle"));
+  ConstantInt* resourceClass = dyn_cast<ConstantInt>(createHandle->getArgOperand(1));
+  if( (nullptr != resourceClass) && (resourceClass->getZExtValue() == 1) )
+    return true;
+
+  return false;
+}
+
+// Rematerialization logic of future compilation should imply basic to moderate optimizations wrt live values.
+bool LiveValueAnalysis::canRemat(Instruction* inst) {
+  if( isa<GetElementPtrInst>( inst ) )
+    return true;
+  else if( isa<AllocaInst>( inst ) )
+    return true;
+  else if( isa<LoadInst>( inst ) )
+    return true;
+
+  if( CallInst* call = dyn_cast<CallInst>(inst) ) {
+    if( call->getCalledFunction()->getName().startswith( "dx.op.bufferLoad" ) && !usesUavResource( call ) )
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.textureLoad") && !usesUavResource(call))
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.rawBufferLoad") && !usesUavResource(call))
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.createHandleForLib"))
+      return true;
+  }
+
+  return false;
+}
+
+// Full path locations can be complicated and some projects have strange behavior. Customize accordingly.
+std::string LiveValueAnalysis::formatSourceLocation( const std::string& m_rootPath, const std::string& Location, int lineNumber )
+{
+  std::string fullPath = "";
+
+  // Only add the root path if one does not already exist for Location.
+  if( Location.find( ':' ) == StringRef::npos )
```

Comment:
Changed to std::string::npos.
As for Linux, I think it could be useful but I'm not able to test it on that platform myself. Also, not sure how much other code would have to be changed. Does DXC have an OS abstraction layer or utilities to help here?  Or is there is even a use case for Linux?